rust/src/lib.rs

```diff
@@ -65,44 +77,74 @@ impl CloudFormationParams {
 
 impl fmt::Display for CloudFormationParams {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(
+        writeln!(
             f,
             "bucket: {}\nkey: {}",
             self.bucket_name,
             match &self.key_arn {
-                None => "None".to_string(),
-                Some(k) => k.to_string(),
+                None => "None",
+                Some(k) => k,
             }
         )
     }
 }
 
+impl Meta {
+    pub fn new_json(algorithm: &str, nonce: &[u8]) -> serde_json::Result<String> {
+        let meta = Meta {
+            alg: algorithm.to_owned(),
+            nonce: base64::engine::general_purpose::STANDARD.encode(nonce),
+        };
+
+        serde_json::to_string(&meta)
+    }
+}
+
+impl S3DataKeys {
+    pub fn new(name: &str) -> S3DataKeys {
+        S3DataKeys {
+            key: format!("{name}.key"),
+            cipher: format!("{name}.aesgcm.encrypted"),
+            meta: format!("{name}.meta"),
+        }
+    }
+
+    pub fn as_array(&self) -> [&str; 3] {
+        [&self.key, &self.cipher, &self.meta]
+    }
+}
+
+impl fmt::Display for Vault {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "region: {}\n{}", self.region, self.cloudformation_params)
+    }
+}
+
 impl Vault {
     pub async fn new(
         vault_stack: Option<&str>,
         region_opt: Option<&str>,
     ) -> Result<Vault, VaultError> {
         let config = aws_config::from_env()
-            .region(get_region_provider(region_opt))
+            .region(Self::get_region_provider(region_opt))
             .load()
             .await;
 
         // Check env variables directly in case library is not used through the CLI.
         // These are also handled in the CLI so they are documented in the CLI help.
-        let vault_stack_from_env = get_env_variable("VAULT_STACK");
-        let vault_bucket_from_env = get_env_variable("VAULT_BUCKET");
-        let vault_key_from_env = get_env_variable("VAULT_KEY");
+        let vault_stack_from_env = Self::get_env_variable("VAULT_STACK");
+        let vault_bucket_from_env = Self::get_env_variable("VAULT_BUCKET");
+        let vault_key_from_env = Self::get_env_variable("VAULT_KEY");
 
         let cloudformation_params = match (vault_bucket_from_env, vault_key_from_env) {
-            (Some(bucket), Some(key)) => {
-                CloudFormationParams::from(bucket.as_str(), Some(key.as_str()))
-            }
+            (Some(bucket), Some(key)) => CloudFormationParams::new(bucket, Some(key)),
```

Comment:
bucket and key can be moved to `CloudFormationParams` since they are not used after this